dev/com.ibm.ws.microprofile.faulttolerance.3.0.metrics/src/com/ibm/ws/microprofile/faulttolerance/metrics/integration/MetricRecorder30Impl.java

```diff
@@ -0,0 +1,469 @@
+import com.ibm.ws.microprofile.faulttolerance.spi.TimeoutPolicy;
+
+/**
+ *
+ */
+public class MetricRecorder30Impl implements MetricRecorder {
+
+    private static final EnumMap<RetryResultCategory, Tag> RETRY_RESULT_TAGS = new EnumMap<>(RetryResultCategory.class);
+    private static final EnumMap<RetriesOccurred, Tag> RETRIES_OCCURRED_TAGS = new EnumMap<>(RetriesOccurred.class);
+    static {
+        RETRY_RESULT_TAGS.put(NO_EXCEPTION, new Tag("retryResult", "valueReturned"));
+        RETRY_RESULT_TAGS.put(EXCEPTION_IN_ABORT_ON, new Tag("retryResult", "exceptionNotRetryable"));
+        RETRY_RESULT_TAGS.put(EXCEPTION_NOT_IN_RETRY_ON, new Tag("retryResult", "exceptionNotRetryable"));
+        RETRY_RESULT_TAGS.put(MAX_DURATION_REACHED, new Tag("retryResult", "maxDurationReached"));
+        RETRY_RESULT_TAGS.put(MAX_RETRIES_REACHED, new Tag("retryResult", "maxRetriesReached"));
+
+        RETRIES_OCCURRED_TAGS.put(WITH_RETRIES, new Tag("retried", "true"));
+        RETRIES_OCCURRED_TAGS.put(NO_RETRIES, new Tag("retried", "false"));
+    }
+
+    private final EnumMap<FallbackOccurred, Counter> invocationSuccessCounter;
+    private final EnumMap<FallbackOccurred, Counter> invocationFailedCounter;
+    private final EnumMap<RetryResultCategory, EnumMap<RetriesOccurred, Counter>> retryCallsCounter;
+    private final Counter retryRetriesCounter;
+    private final Histogram timeoutDurationHistogram;
+    private final Counter timeoutTrueCalls;
+    private final Counter timeoutFalseCalls;
+    private final Counter circuitBreakerCallsSuccessCounter;
+    private final Counter circuitBreakerCallsFailureCounter;
+    private final Counter circuitBreakerCallsOpenCounter;
+    @SuppressWarnings("unused")
+    private final Gauge<Long> circuitBreakerOpenTime;
+    @SuppressWarnings("unused")
+    private final Gauge<Long> circuitBreakerHalfOpenTime;
+    @SuppressWarnings("unused")
+    private final Gauge<Long> circuitBreakerClosedTime;
+    private final Counter circuitBreakerTimesOpenedCounter;
+    @SuppressWarnings("unused")
+    private final Gauge<Long> bulkheadConcurrentExecutions;
+    private final Counter bulkheadRejectionsCounter;
+    private final Counter bulkheadAcceptedCounter;
+    private final Histogram bulkheadExecutionDuration;
+    @SuppressWarnings("unused")
+    private final Gauge<Long> bulkheadQueuePopulation;
+    private final Histogram bulkheadQueueWaitTimeHistogram;
+    private LongSupplier concurrentExecutionCountSupplier = null;
+    private LongSupplier queuePopulationSupplier = null;
+    private long openNanos;
+    private long halfOpenNanos;
+    private long closedNanos;
+
+    private enum CircuitBreakerState {
+        CLOSED,
+        HALF_OPEN,
+        OPEN
+    }
+
+    public MetricRecorder30Impl(String methodName, MetricRegistry registry, RetryPolicy retryPolicy, CircuitBreakerPolicy circuitBreakerPolicy, TimeoutPolicy timeoutPolicy,
```

Comment:
This class is kinda hard to read in a github diff ... do you think it needs more doc?